--- a/config.def.h
+++ b/config.def.h
@@ -65,12 +65,19 @@
 static char dmenumon[2] = "0"; /* component of dmenucmd, manipulated in spawn() */
 static const char *dmenucmd[] = { "dmenu_run", "-m", dmenumon, "-fn", dmenufont, "-nb", col_gray1, "-nf", col_gray3, "-sb", col_cyan, "-sf", col_gray4, NULL };
 static const char *termcmd[]  = { "st", NULL };
+
+#include "shift-tools.c"
 #include "movestack.c"
+
 static Key keys[] = {
 	/* modifier                     key        function        argument */
+	{ MODKEY,                       XK_i,      shiftview,      { .i = -1 } },
+	{ MODKEY,                       XK_o,      shiftview,      { .i = +1 } },
+	{ MODKEY|ControlMask,           XK_i,      shiftviewclients, { .i = -1 } },
+	{ MODKEY|ControlMask,           XK_o,      shiftviewclients, { .i = +1 } },
+	{ MODKEY|ShiftMask,             XK_i,      shiftboth,      { .i = -1 }	},
+	{ MODKEY|ShiftMask,             XK_o,      shiftboth,      { .i = +1 }	},
 	{ MODKEY,                       XK_p,      spawn,          {.v = dmenucmd } },
 	{ MODKEY|ShiftMask,             XK_Return, spawn,          {.v = termcmd } },
 	{ MODKEY,                       XK_b,      togglebar,      {0} },
--- a/dwm.c
+++ b/dwm.c
@@ -194,6 +194,7 @@
 static void sendmon(Client *c, Monitor *m);
 static void setclientstate(Client *c, long state);
 static void setfocus(Client *c);
+static void setfullscreen(Client *c, int fullscreen);
 static void setlayout(const Arg *arg);
 static void setmfact(const Arg *arg);
 static void setup(void);
--- a/shift-tools.c
+++ b/shift-tools.c
@@ -0,0 +1,114 @@
+void
+shiftview(const Arg *arg) {
+	Arg shifted;
+
+	if(arg->i > 0) // left circular shift
+		shifted.ui = (selmon->tagset[selmon->seltags] << arg->i)
+		   | (selmon->tagset[selmon->seltags] >> (LENGTH(tags) - arg->i));
+	else // right circular shift
+		shifted.ui = selmon->tagset[selmon->seltags] >> (- arg->i)
+		   | selmon->tagset[selmon->seltags] << (LENGTH(tags) + arg->i);
+
+	view(&shifted);
+}
+
+void
+shiftviewclients(const Arg *arg)
+{
+	Arg shifted;
+	Client *c;
+	unsigned int tagmask = 0;
+
+	if(arg->i > 0) // left circular shift
+		shifted.ui = (selmon->tagset[selmon->seltags] << arg->i)
+		   | (selmon->tagset[selmon->seltags] >> (LENGTH(tags) - arg->i));
+	else // right circular shift
+		shifted.ui = selmon->tagset[selmon->seltags] >> (- arg->i)
+		   | selmon->tagset[selmon->seltags] << (LENGTH(tags) + arg->i);
+
+	for (c = selmon->clients; c; c = c->next) {
+		if (c->tags & selmon->tagset[selmon->seltags])
+			c->tags = shifted.ui;
+	}
+
+	focus(NULL);
+	arrange(selmon);
+	view(&shifted);
+}
+
+void
+shiftboth(const Arg *arg)
+{
+	Arg shifted;
+
+	if(arg->i > 0) // left circular shift
+		shifted.ui = (selmon->tagset[selmon->seltags] << arg->i)
+		   | (selmon->tagset[selmon->seltags] >> (LENGTH(tags) - arg->i));
+	else // right circular shift
+		shifted.ui = selmon->tagset[selmon->seltags] >> (- arg->i)
+		   | selmon->tagset[selmon->seltags] << (LENGTH(tags) + arg->i);
+
+	selmon->sel->tags = shifted.ui;
+	focus(selmon->sel);
+	arrange(selmon);
+	view(&shifted);
+}
+
+
+void
+shifttag(const Arg *arg) {
+	Arg shifted;
+
+	if(arg->i > 0) // left circular shift
+		shifted.ui = (selmon->tagset[selmon->seltags] << arg->i)
+		   | (selmon->tagset[selmon->seltags] >> (LENGTH(tags) - arg->i));
+	else // right circular shift
+		shifted.ui = selmon->tagset[selmon->seltags] >> (- arg->i)
+		   | selmon->tagset[selmon->seltags] << (LENGTH(tags) + arg->i);
+
+	tag(&shifted);
+}
+
+void
+shifttagclients(const Arg *arg)
+{
+	Arg shifted;
+	Client *c;
+	unsigned int tagmask = 0;
+
+	if(arg->i > 0) // left circular shift
+		shifted.ui = (selmon->tagset[selmon->seltags] << arg->i)
+		   | (selmon->tagset[selmon->seltags] >> (LENGTH(tags) - arg->i));
+	else // right circular shift
+		shifted.ui = selmon->tagset[selmon->seltags] >> (- arg->i)
+		   | selmon->tagset[selmon->seltags] << (LENGTH(tags) + arg->i);
+
+	for (c = selmon->clients; c; c = c->next) {
+		if (c->tags & selmon->tagset[selmon->seltags]) {
+			c->tags = (c->tags & ~selmon->tagset[selmon->seltags]) | shifted.ui;
+			tagmask |= shifted.ui;
+		}
+	}
+
+	if (tagmask) {
+		selmon->tagset[selmon->seltags] = tagmask;
+		view(&shifted);
+	}
+}
+
+void
+shifttagboth(const Arg *arg)
+{
+	Arg shifted;
+
+	if(arg->i > 0) // left circular shift
+		shifted.ui = (selmon->tagset[selmon->seltags] << arg->i)
+		   | (selmon->tagset[selmon->seltags] >> (LENGTH(tags) - arg->i));
+	else // right circular shift
+		shifted.ui = selmon->tagset[selmon->seltags] >> (- arg->i)
+		   | selmon->tagset[selmon->seltags] << (LENGTH(tags) + arg->i);
+
+	selmon->sel->tags = (selmon->sel->tags & ~selmon->tagset[selmon->seltags]) | shifted.ui;
+	arrange(selmon);
+	view(&shifted);
+}
